File: backend/app/schemas/books.py
"""Book Engine response schemas (spec 14)."""

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field


class SubjectOut(BaseModel):
    id: int
    name: str
    grade: int
    track: str
    type: str  # noqa: A003 — matches spec 04 column name


class BookOut(BaseModel):
    id: int
    stable_key: str
    title: str
    publisher: str
    edition: str
    config_version: int
    grade: int
    track: str
    subject: SubjectOut
    active: bool
    activated_at: datetime | None
    node_count: int
    test_set_count: int
    question_count: int


# Detail currently equals list item (tree lives in a dedicated endpoint).
BookDetailOut = BookOut


class TestSetSummaryOut(BaseModel):
    id: int
    title: str
    test_type: str
    node_id: int | None
    question_count: int
    meta: dict[str, Any] = Field(default_factory=dict)


class TreeNodeOut(BaseModel):
    id: int
    parent_id: int | None
    node_type: str
    title: str
    code: str
    order_index: int
    is_leaf: bool
    meta: dict[str, Any] = Field(default_factory=dict)
    test_sets: list[TestSetSummaryOut] = Field(default_factory=list)
    children: list["TreeNodeOut"] = Field(default_factory=list)


TreeNodeOut.model_rebuild()


class BookTreeOut(BaseModel):
    book_id: int
    nodes: list[TreeNodeOut]


class ChildNodeOut(BaseModel):
    id: int
    parent_id: int | None
    node_type: str
    title: str
    code: str
    order_index: int
    is_leaf: bool
    meta: dict[str, Any] = Field(default_factory=dict)
    test_sets: list[TestSetSummaryOut] = Field(default_factory=list)


class NodeChildrenOut(BaseModel):
    node_id: int
    children: list[ChildNodeOut]


class ActivationOut(BaseModel):
    user_id: int
    book_id: int
    active: bool
    activated_at: datetime


class ImportOut(BaseModel):
    status: Literal["imported", "unchanged"]
    book_id: int
    stable_key: str
    node_count: int
    test_set_count: int
    question_count: int

File: frontend/src/types/books.ts
/** Book Engine types — mirrors backend schemas/books.py. */

export interface Subject {
  id: number;
  name: string;
  grade: number;
  track: string;
  type: string;
}

export interface Book {
  id: number;
  stable_key: string;
  title: string;
  publisher: string;
  edition: string;
  config_version: number;
  grade: number;
  track: string;
  subject: Subject;
  active: boolean;
  activated_at: string | null;
  node_count: number;
  test_set_count: number;
  question_count: number;
}

export interface TestSetSummary {
  id: number;
  title: string;
  test_type: string;
  node_id: number | null;
  question_count: number;
  meta: Record<string, unknown>;
}

export interface TreeNode {
  id: number;
  parent_id: number | null;
  node_type: string;
  title: string;
  code: string;
  order_index: number;
  is_leaf: boolean;
  meta: Record<string, unknown>;
  test_sets: TestSetSummary[];
  children: TreeNode[];
}

export interface BookTree {
  book_id: number;
  nodes: TreeNode[];
}

export interface Activation {
  user_id: number;
  book_id: number;
  active: boolean;
  activated_at: string;
}

export interface ImportResult {
  status: "imported" | "unchanged";
  book_id: number;
  stable_key: string;
  node_count: number;
  test_set_count: number;
  question_count: number;
}

export interface ConfigIssue {
  path: string;
  message: string;
}
